drivers/i2c/i2c_cc32xx.c

```diff
@@ -0,0 +1,417 @@
+/*
+ * Copyright (c) 2017, Texas Instruments Incorporated
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* The logic here is adapted from SimpleLink SDK's I2CCC32XX.c module. */
+
+#include <kernel.h>
+#include <errno.h>
+#include <i2c.h>
+#include <soc.h>
+#include "i2c-priv.h"
+
+/* Driverlib includes */
+#include <inc/hw_memmap.h>
+#include <inc/hw_common_reg.h>
+#include <driverlib/rom.h>
+#include <driverlib/rom_map.h>
+#include <driverlib/i2c.h>
+
+#define I2C_MASTER_CMD_BURST_RECEIVE_START_NACK	 I2C_MASTER_CMD_BURST_SEND_START
+#define I2C_MASTER_CMD_BURST_RECEIVE_STOP \
+	I2C_MASTER_CMD_BURST_RECEIVE_ERROR_STOP
+#define I2C_MASTER_CMD_BURST_RECEIVE_CONT_NACK	 I2C_MASTER_CMD_BURST_SEND_CONT
+
+#define I2C_SEM_MASK \
+	COMMON_REG_I2C_Properties_Register_I2C_Properties_Register_M
+#define I2C_SEM_TAKE \
+	COMMON_REG_I2C_Properties_Register_I2C_Properties_Register_S
+
+#define IS_I2C_MSG_WRITE(flags) ((flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE)
+
+/* Remove if/when this shows up in i2c.h: */
+#define I2C_SPEED_GET(cfg)		(((cfg) & I2C_SPEED_MASK) \
+						>> I2C_SPEED_SHIFT)
+
+#define DEV_CFG(dev) \
+	((const struct i2c_cc32xx_config *const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct i2c_cc32xx_data *const)(dev)->driver_data)
+#define DEV_BASE(dev) \
+	((DEV_CFG(dev))->base)
+
+
+enum i2c_cc32xx_state {
+	/* I2C is idle, and not performing a transaction */
+	I2C_CC32XX_IDLE_MODE = 0,
+	/* I2C is currently performing a write operation */
+	I2C_CC32XX_WRITE_MODE,
+	/* I2C is currently performing a read operation */
+	I2C_CC32XX_READ_MODE,
+	/* I2C error has occurred */
+	I2C_CC32XX_ERROR = 0xFF
+};
+
+struct i2c_cc32xx_config {
+	u32_t base;
+	u32_t bitrate;
+	unsigned int exception_no;
+};
+
+struct i2c_cc32xx_data {
+	struct k_sem mutex;
+	struct k_sem transfer_complete;
+
+	volatile enum i2c_cc32xx_state state;
```

Comment:
It seems like the state is not really used in the driver code, there are no state machine like dependencies. We only differentiate between normal operation and an error condition. We could replace this enum with a single flag.

Reply:
Actually, the isr has a state machine which determines the next path depending on how the I2C transfer was primed: for read or write.  The distinction of read vs write could have been determined from the read or write index variables, but we coalesced those into one.   It seems the state variable is still needed here.

Reply:
You are right, I completely overlooked the `switch (data->state)` statement in the ISR. That said, I think we should have a closer look at the ISR. Your current implementation may work well but has some weaknesses. 

The typical way to implement an ISR would be to react to specific status flags ISR was invoked with. I.e. the conditional statement should be built around value of I2CMMIS "I2C Master Masked Interrupt Status" register (I gave only a brief look into the datasheet, but I think I got it right). Often more than one flag can be activated simultaneously and checking the flags directly is the most reliable way to determine the reason ISR was invoked. If you attempt to infer the cause of ISR from other sources, e.g. the data->state variable there is a chance you get it wrong. What if some other interrupt mask was enabled which was not taken into account? Some of the other interrupts may happen only under specific conditions. Such errors will not easily be caught under basic testing conditions.

Using data->state variable to determine ISR cause makes it also difficult to extend the driver in the future. A direct dependency on interrupt status is much easier to work with when new features are added.

Reply:
That seems reasonable.  Will need to look at modifying the logic in the next iteration.

Reply:
@mnkp, @galak, finally got back to this task, and made recommended changes to the ISR.

The ISR now is (mainly) driven by the interrupt status bits.  
Since in this driver, FIFOs are not directly enabled and used, I found no interrupt registers telling whether the data transfer was primed as a read or a write, so I kept those state variables only for that purpose.  Everything else is driven by the ISR flags.